File: 백준/Gold/2098. 외판원 순회/외판원 순회.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {
    private static final int INF = (int) 1e9;
    static StringTokenizer st;
    static int[][] edges, memo;
    static int n, end;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        end = (1 << n) - 1;

        edges = new int[n][n];

        for (int r = 0; r < n; r++) {
            st = new StringTokenizer(br.readLine());

            for (int c = 0; c < n; c++) {
                edges[r][c] = nextInt();
            }
        }

        memo = new int[end + 1][n];

        System.out.println(travel(0, 1));
    }

    static int travel(int u, int visit) {
        if (visit == end) {
            if (edges[u][0] == 0) {
                return INF;
            }

            return edges[u][0];
        }

        if (memo[visit][u] > 0) {
            return memo[visit][u];
        }

        int ret = INF;

        for (int v = 1; v < n; v++) {
            if (edges[u][v] == 0) continue;

            int mask = 1 << v;

            if ((visit & mask) == mask) continue;

            ret = Math.min(ret, edges[u][v] + travel(v, visit | mask));
        }

        memo[visit][u] = ret;

        return ret;
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }
}

File: 백준/Gold/2098. 외판원 순회/외판원 순회.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine().toInt()
    val costs = Array(n) { IntArray(n) }

    repeat(n) { r ->
        val line = StringTokenizer(readLine())
        for (c in 0 until n) {
            costs[r][c] = line.nextToken().toInt()
        }
    }

    val tsp = TSP(n, costs)
    tsp.tsp()
    print(tsp.getMin())
}

class TSP(val n: Int, val costs: Array<IntArray>) {
    val memo = Array(1 shl n) { IntArray(n) { Int.MAX_VALUE } }

    fun tsp() {
        travel(1, 0, 0)
        val last = memo.lastIndex

        for (i in 0 until n) {
            if (costs[i][0] == 0) {
                memo[last][i] = Int.MAX_VALUE
            }
            else if (memo[last][i] != Int.MAX_VALUE) {
                memo[last][i] += costs[i][0]
            }
        }
    }

    fun travel(mask: Int, u: Int, cost: Int) {
        if (mask == memo.lastIndex) {
            if (cost < memo[mask][u]) {
                memo[mask][u] = cost
            }
            return
        }
        for (v in 0 until n) {
            val bit = 1 shl v

            if (mask and bit == bit || costs[u][v] == 0) {
                continue
            }
            val nextMask = mask or bit
            val nextCost = cost + costs[u][v]

            if (nextCost < memo[nextMask][v]) {
                memo[nextMask][v] = nextCost
                travel(nextMask, v, nextCost)
            }
        }
    }

    fun getMin() = memo[memo.lastIndex].min()
}